### ISSUE TITLE
setup & build configuration

### ISSUE THREAD

**Issue by Thunder07** [jpd002/Play-#506]
the old 'issue' seems to have disappeared, probably because fadi was banned.

> Maybe we could add the generated project in the repo to make the build setup process easier?

but isn't that what happening now? on windows you've the sln project file, or osx you've Xcode project and linux users are two lines from installing and using cmake, cmake would just generate the same files on each platform, the point of cmake is to have 1 build file that can be modified for all platforms. if there is an issue its probably about setting up the build (mostly osx because as you mentioned boost can be a pain to setup) and as you suggested just having a script that works for all platform.
I also agree with @Orphis cmake is common enough it shouldn't put anyone that wants to help.

> boost also makes the setup slightly complicated on macOS and iOS because you need to build the frameworks.

is there a reason it has to be built as a framework? a while back i installed boost using brew and managed to get Play to build, If I'm not mistaken they were a regular library build, if so this solves that problem no?

I'll setup fresh builds when i get the chance and see how it works out.(edit just built play with boost from brew, instructions in post below)

> Maybe we would just need minimal documentation to list package dependencies.

> sudo apt-get install python-dev libbz2-dev libalut-dev libmysqlclient-dev libglew-dev libicu-dev cmake git -y
> sudo apt-get install libboost-filesystem-dev libboost-chrono-dev

this was what I needed when I working on the linux build, i guess gcc also need to be added to that

in retrospect I've never found an issue setting up the build, I think it's just matter of documenting the steps to get it up and running.

@Phatcat I was thinking of the tablet, but none the less there shouldnt be any reason to root your device, apk is available on the store (not sure if it will display on TV store) and you can also side load it.

> what's the need for cmake? linux/unix builds? Does this project have any VS project files associated

yes, yes.
VS project for windows, cmake to generate makefile for ubuntu/linux builds and Xcode project for osx.

**Comment by Thunder07** [comment]
ubuntu(haven't tested building since a month ago):
install qt [5.6.*](https://www.qt.io/download-open-source/#section-2) or you can install it using `sudo apt-get install qt5-default` but that's version 5.5 and will spam the terminal with warning msgs when you start Play-Ui
```
sudo apt-get install python-dev libbz2-dev libalut-dev libmysqlclient-dev libglew-dev libicu-dev cmake git -y
sudo apt-get install libboost-filesystem-dev libboost-chrono-dev -y

git clone https://github.com/jpd002/Play-Build.git Play-Build
cd Play-Build
git submodule update --init --recursive
git submodule foreach git pull origin master

cd Play/build_unix
sh ./build.sh
```

OSX with small modifications:

```
brew install boost

git clone https://github.com/jpd002/Play-Build.git Play-Build
cd Play-Build
git submodule update --init --recursive
git submodule foreach git pull origin master
```

open Play in Xcode, modified both Play's and Framework's GeneralSettings.xconfig:
`BOOST_FRAMEWORK_PATH = /usr/local/Cellar/boost/1.60.0_1` #you can probably rename the variable too.
`HEADER_SEARCH_PATHS = $(BOOST_FRAMEWORK_PATH)/include` instead of framework path
removed `FRAMEWORK_SEARCH_PATHS = $(BOOST_FRAMEWORK_PATH)`

then open go the project file of both Play and Framework, Build Settings->Search Paths and change Framework/Header/Library Search Paths and append these as per above changes.

Lastly go to Play's project file Build Settings->Linking:
Other Linker Flags: `-ObjC -lboost_filesystem -lboost_system -lboost_chrono`

then build.

you can also build it using terminal.
```
cd Play/build_macosx
xcodebuild -configuration Debug
```

Windows: looking at you appvoyer it seems straight forward.

now android is the only build that requires building boost directly. I'll update this post later with instruction for it.

**Comment by Orphis** [comment]
The problem with Boost from homebrew is that you don't get iOS binaries. 
The proper way to use Boost in such a project is to download the sources from Cmake, define static libraries for the compiled modules you use and just build it. 
It's very easy to do, and absolutely painless. It will work for all platforms too. I did that at my previous company and we had software for even more platforms.

Many people asked for that, so I should really work on a small open source project doing exactly that.

**Comment by Thunder07** [comment]
I think the issue with boost is that its being compiled as a framework, which is creating unnecessary work, again unless I'm missing the point of framework libraries.

**Comment by Orphis** [comment]
@Thunder07 No, that won't work since there are tons of architectures, compiler options, instrumentations that would require a rebuild, it's hard to pick automatically which prebuilt binary to use (and also to cover all use case).
That's why it's just easier to integrate it directly in your project's build so it inherits everything automatically.

Also, it would cover other platforms like Windows, Android or Linux. One solution for everything!

**Comment by Phatcat** [comment]
Well, it's not like I would be able to get my own test builds from the google app store :p

Hmm.. I still prefer buildbot over appvoyer/travis.

Dolphin has it set up in such a neat way that if you PR a change buildbot will show each build it will do as a different github integrated test directly in the PR (look at any one of the open PRs on the dolphin emulator project); this way the developer and everyone else interested in the change have easy, direct access to an already build app ready to run (once it's done, of course) for each and every single platform they do, which also allows developers to develop and test even when a build environment is not available to them.
To make sure it's not abused they have a simple check in the beginning if you're a trusted developer or not, if not then buildbot isn't started at all; then a trusted developer has to manually check positive the first check on that PR before it runs the checks (builds the different release/debug versions)

Someone got banned? how come? also how often does that nuke an entire associated issue? on a repo that still stands? O.o

**Comment by Phatcat** [comment]
Also if you're going to compile a bunch of info on how to set up a build environment on different work environments and building the project for different platforms, how about compiling it into a wiki page so it can be adjusted, updated and modified, and so that it won't get lost in time? =)

**Comment by Orphis** [comment]
@Phatcat That's not how it even remotely works.
You need to have your own servers to run buildbot and provide the downloads. Dolphin has enough visitors on their website to cover the cost of servers, so they can do it. I don't think that's the case for this project.
On the other hand, Travis and Appveyor are free. And nothing prevents you from uploading the resulting binary anywhere for download later.

**Comment by Thunder07** [comment]
@Phatcat he has a history of spamming requests and according to the reddit post he made, he spammed several different emulators repo with the same request to help this one, it was only a mater of time really.

and thats the plan but 1st I'll try to get a setup to work for each platform (excluding windows) before doing so, just so I'm just so I want to see I want to get a build for iOS(no iPhone to test)/OSX down with source built boost without framework, if @jpd002 doesn't have an issue with that, which will require a PR to go with it as well.

**Comment by Phatcat** [comment]
Orphis, I did not state anything that is directly false... Yes.. it is how it works, specifically for dolphin, even if only 'remotely'.. ;)  I did consider the posibility that you did need your own server, but I was under the slight impression that there were free solutions out there for open source projects, just like both travis-ci and appvoyer can provide their services for free to open source projects, like github can provide it for free to us as well, etc..

Also, I still prefer the system that buildbot provides, even if you have to provide the hardware yourself, but I can understand how it's not an option in this case, no need to be so aggressive about it .. =)

He must've spammed quite a bit to have gotten the ban hammer to the point it nukes an entire issue instead of just wiping it for all of his existance.

**Comment by Orphis** [comment]
@Phatcat Yet, you can't compare a custom tailored solution for a very active team to this project.
There are some free solutions, but since that will require server time, buildbot is never the software powering those. To be honest, it's probably the least friendly software of all the CI tools around. Because you have a good experience with it from Dolphin doesn't mean anyone else is having much fun using it in other projects!

Free infra is much better when you have a small project like this, you don't need the complexity of something you have to manage and attend to often.

**Comment by Phatcat** [comment]
That's a fair point, there could be a lot more maintenance, setup, cost and general headache involved in effectively keeping specifically 'buildbot' up and working. I'm not denying that.. the point was more on the end result; having all the different platform builds as checks on all pull requests and if successful, have each different build available for download. I was really not having a go or making demands or anything

**Comment by Thunder07** [comment]
I'm working to compile for each platform so i can great a build guide, when I was trying to build boost with the script provided for osx/ios but it kept failing, however using a newer script fixed the issue, is anyone else having this issue?

**Comment by Orphis** [comment]
@Phatcat Travis and Appveyor can actually review pull requests. That's not something easy to do with Buildbot installations since you usually don't want to build and run untrusted code on your workers. Publishing the builds for download is a different matter altogether and easily solved (kinda).

@Thunder07 I'd say don't bother. There should be a focus on unifying the build scripts with CMake properly first.
It's unacceptable to have so many different set of build scripts for a cross platform project in (almost) 2017.

Then, you will be able to use my project that I just pushed to build Boost from source directly with the rest of your program with CMake: https://github.com/Orphis/boost-cmake . It should just work as it is an adaptation of what I used at my previous employer and is used by over 100 million software installs on all platforms.

**Comment by Thunder07** [comment]
@Orphis I've built this project on all 3 desktop platform with CMake before, at the time my aim was to get them all running the same qt ui, but I dropped it since there was some bugs with Windows/OSX builds. I can probably go about using cmake to atleast get the libraries built while the UI can stay OS specific for now.

the question is, will that build boost for iOS? if so I can probably have something ready in a few.

**Comment by Orphis** [comment]
It will. I supports also universal builds. 
That's what Spotify has been using for a while now to integrate Boost. 
Just do add my project as a submodule, add_subdirectory(boost-cmake) and do target_link_libraries (Foo Boost::filesystem Boost::system). No need to specify the include directories if you link with the libraries in the right targets!

**Comment by Orphis** [comment]
One more thing, it is using Boost 1.62 and will link it statically. So on Linux or Android, you need to attach on the right libraries or you will have link order issues. 
Feel free to create a PR with your CMake work and I can review that and suggest modern techniques to make it easier to work with.

**Comment by Thunder07** [comment]
@Orphis I've them on my repos but I'm not sure if that's the way I should be adding boost into the current cmake, but it's the only way i figured without dependencies clash.

git clone -b cmake_test https://github.com/Thunder07/Play-.git Play
git clone -b cmake_test https://github.com/Thunder07/Play--Framework.git Framework
git clone https://github.com/jpd002/Play--CodeGen.git CodeGen
git clone -b cmake_test https://github.com/Thunder07/Play-Dependencies.git Dependencies

**Comment by Thunder07** [comment]
using @Orphis's boost-cmake removed the hurdle step of setting up boost, so it made it much easier to update the osx and linux build to completely build start to end using cmake.
as of now ubuntu is using qt ui, while osx is using xib, but I'll probably add a variable to allow the user to also build the qt ui for osx. if i get the time tomorrow, i might look into moving windows over to cmake as well.

as for any dependencies, they can be installed with brew/apt as instructed in the 2nd post above.

Ubuntu/OSX:
```
git clone -b master_cmake_osx https://github.com/Thunder07/Play-.git Play
git clone -b cmake-boost https://github.com/Thunder07/Play--Framework.git Framework
git clone https://github.com/jpd002/Play--CodeGen.git CodeGen
git clone -b cmake-boost https://github.com/Thunder07/Play-Dependencies.git Dependencies

cd Play/build_unix
mkdir build
cd build
```

OSX:
```
cmake .. -GXcode
cmake --build . --config Release
```

Ubuntu:
```
#if you installed qt using apt you can remove or ignore CMAKE_PREFIX_PATH
#if you installed qt using the qt installer you need to update CMAKE_PREFIX_PATH to reflect the correct path
cmake .. -DCMAKE_PREFIX_PATH=~/Qt5.6.0/5.6/gcc_64/
make
```

**Comment by Thunder07** [comment]
Windows:
```
git clone -b windows_cmake_test https://github.com/Thunder07/Play-.git Play
git clone -b cmake_win_test https://github.com/Thunder07/Play--Framework.git Framework
git clone https://github.com/jpd002/Play--CodeGen.git CodeGen
git clone -b cmake_win_test https://github.com/Thunder07/Play-Dependencies.git Dependencies
#need to make sure boost-cmake is from my repo for now, as their is linking issue on windows
git clone -b cmake_windows https://github.com/Thunder07/Nuanceur.git

cd Play/build_unix
mkdir build
cd build

cmake .. -G"Visual Studio 14 2015 Win64" -DGLEW_LOCATION="C:/Program Files/glew"
#had to use the glew32.lib (none static one to build correctly) which meant including glew32.dll in the exe folder to run it
cmake --build . --config Release
```
the for the dependencies mysql client (mysql connector c), open al sdk, directx sdk, of which i downloaded the installer. glew i downloaded the zip and extracted it to that location.

**Comment by jpd002** [comment]
Thanks, I'll give this a try and give you my feedback.

**Comment by Orphis** [comment]
There are build issues because you mix sources of Boost. 
There is a feature called auto linking that I disable in my build and propagate the setting for when you use it properly. 
But you don't, so other parts of the emulator will try to use it and search for Boost with another name that is expected for autolinking.

**Comment by Thunder07** [comment]
How am I mixing sources? This was done on a fresh windows install with a fresh build of play, aka no other build of play or boost exist here. 
The issue is that auto linking isn't disabled in your build for whatever reason, I see the flag but it's not working, it's either not set correctly or the preprocessor is also needed to disable it.

**Comment by Orphis** [comment]
You have targets using Boost but not through the CMake targets that will propagate the include directory. 
Remove the include_directories ( BOOST_SOURCE) to find them. You really should not be doing that.

**Comment by Thunder07** [comment]
> Remove the include_directories ( BOOST_SOURCE) to find them. You really should not be doing that.

I was meant to remove and test this with `Boost::boost`, but didn't get the time to...
but this is set by your script itself and shouldn't effect the linking, but the headers no?
the errors i got was an error linking to libboost_filesystem-msv140-mt.lib even though I can see it trying to link to libboost_filesystem.lib

at anycase I'll try that later when I get the chance.

**Comment by Orphis** [comment]
No, it's exactly as I say.
Some other component is using Boost through that include_directories(), so it doesn't get the define to prevent it from using autolinking. Some other library is using the target Boost::filesystem properly, so it tries to get both the one from CMake and the one the headers tell it to find.

That's the problem with global effects like the ones from include_directories or add_definitions. It's better to use properly defined local dependencies.

**Comment by Thunder07** [comment]
yup, that did the trick, could have saved myself few hours of heartache if I've just done it.

**Comment by jpd002** [comment]
I've been playing with this a little and I think it's going in the right direction. Since we're trying to make the setup process as easy as possible, I think there's some more things we should consider.

Right now, to build on Windows with @Thunder07's proposal, you need to install GLEW, the OpenAL SDK and MySQL Connector. To fix that, I think we should make separate `CMakeLists.txt` file for every subproject (ie.: opengl, openal, mysql). I was thinking something like this:

```
- Framework
|- build_unix
 |- Framework
  |- CMakeLists.txt
 |- FrameworkOpenAl
  |- CMakeLists.txt
 |- FrameworkOpenGl
  |- CMakeLists.txt
 |- FrameworkMySql
  |- CMakeLists.txt
```
Then the `CMakeLists.txt` file for the Play! project would reference the subprojects as needed (ie.: for Win32, you would need `Framework` and `FrameworkOpenGl`, for Unix you would need `Framework`, `FrameworkOpenGl` and `FrameworkOpenAl`.) This way, you wouldn't need to install OpenAL SDK and MySQL Connector to build on Windows.

Then there's still the requirement to install GLEW on Windows. I think this could probably be circumvented by embedding a snapshot (or submodule) of GLEW's Git repo in `Dependency` and adding a step to build it a bit like what's done for zlib and bzlib.

I'm going to try building for other platforms, but I think it's looking good so far.

**Comment by Orphis** [comment]
Out of curiosity, what is MySQL used for in an emulator?
Traditionally, you recurse in all external dependencies and then in the code for each, you define it or skip it on unsupported platforms. It keeps the external code much simpler.

As for GLEW, this is exactly what is done in other projects, try to find a copy on the system and then fallback to the embedded sources. 

Something that would help too would be to merge the projects. Is there a solid reason to keep them in multiple repositories? Merging them helps to maintain a coherent "workspace" between them all.

**Comment by Thunder07** [comment]
yup, that was going to be my next suggestion after you're happy with the current cmake script.

but for some reason re-reading your post is getting me confused.. maybe its a sight i should go to bed... but why are you adding `mysql` into `framework` while `glew` into `Dependencies`?
are you proposing to sub divide framework in this case?

I've also gotten the android build to work also, however debug builds fine, but most games i've tried throw an exception, while the release build adds `-lrt` flag to the build causing it to fail during linking since that library doesn't exist on android and I'm not sure where the flag is coming from, to make it more odd, after failing re-running the build without a clean would link it correctly.

also what do you think about [this](https://github.com/Thunder07/Play-/blob/uni_cmake/build_unix/CMakeLists.txt#L91), I think having this for all the builds would be beneficial especially if let say, you're building for iOS, OSX and android on the same PC. if so this would need to be added to all the builds.

@Orphis as far as i can tell, it's not being linked to anything, so it's probably not being used and as I mentioned to s1n it's probably made in order to unify the sql code. but mysql... sqlite3 is currently being used on iOS and android to find game's cover and description.

you can clone the entire project through `Play-Build` repo, I think keeping them separate is a design choice to keep the codebase clean, as their is a fine dividing line were everything must go, personally Im for it.

**Comment by jpd002** [comment]
@Orphis Play! doesn't use the MySQL wrapper stuff in Framework. It's only used by [SeventhUmbral](https://github.com/jpd002/SeventhUmbral). I went with the idea of having separate repositories because I thought it would be the best way to share code with my other projects. Integrating those as submodules could be interesting to make it simpler for newcomers I guess. It's something we could look into later.

@Thunder07 I meant, creating a separate `CMakeLists.txt` file for the `mysql` subproject a bit like there is a `FrameworkMysql.vcxproj` in the `build_win32` folder right now. And I meant, adding the actual [GLEW](https://github.com/nigels-com/glew) repo in `Dependencies` to remove the need of installing the library on the user's system if it's not present.

**Comment by Thunder07** [comment]
Ok, I'll see how much i can get done this weekend.

@Orphis android seems to be failing with linking error to -lrt, which shouldn't be there anyway.
after few commenting, i narrowed it to this line https://github.com/Orphis/boost-cmake/blob/master/CMakeLists.txt#L104 any idea whats going on?

**Comment by Orphis** [comment]
I have a fix in the "tests" branch. I'll push it to master later today!

**Comment by Thunder07** [comment]
"tests" seems to be doing the trick

**Comment by Thunder07** [comment]
Progress update, I've cleaned up the commits and should be closer to the end line than the beginning.
few remaining thing:
1) I haven't tested this on windows, but it's missing the directx find code in cmake (quick fix)
2) to move the CMakeLists.txt to the root directory but
3) what to do with current projects, delete?
4) create `CMakeLists.txt` for `tools/PsfPlayer`

I've started my clinical placement, so I'm short for time, so it's likely I'll only be working weekends to get this done.

if you have any comments, feel free to mention them.
```
git clone -b final_cmake https://github.com/Thunder07/Play-.git Play
git clone -b final_cmake https://github.com/Thunder07/Play--Framework.git Framework
git clone -b final_cmake https://github.com/Thunder07/Play--CodeGen.git CodeGen
git clone -b final_cmake https://github.com/Thunder07/Play-Dependencies.git Dependencies
cd Dependencies
git submodule init
git submodule update
# until tests branch of boost-cmake is merged, for testing convenience I'd recommend 
# cd boost-cmake && git checkout tests && git cherry-pick fe127fa78681846d21a194850457381db0a10f18 a68a8b95273196ffe931a6120c9ae6c2e595f4d9 && cd ..
cd ..
git clone -b final_cmake https://github.com/Thunder07/Nuanceur.git
cd Play/build_unix
mkdir build && cd build
#osx works with Xcode and makefile
#windows create VS projects (also makefile i think, but i never tested that)
#unix makefile
#and ninja is multiplatform so all of them (tested on ubuntu)
#android, you need to open android studio and build for there, gradle is not working correctly for me yet
cmake .. #will generate a makefile
cmake .. -GXcode
cmake .. -GNinja
cmake .. -G"Visual Studio 14 2015 Win64"
cmake --build . # you can also decided to open the build project in Xcode/VS and build from there
```

**Comment by iOS4all** [comment]
That's very nice. well after getting your commits to work what's the result should be? I mean can speed up the performance or what exactly? Also what's gradle meaning? also can this works and fixes should be also working on iOS as well?
Thanks.

**Comment by Thunder07** [comment]
@iOS4all fadi you got banned once and you should learn from it.

**Comment by Orphis** [comment]
@Thunder07 No, this is overkill for GLEW. Just have the sources in the same repository really.
You don't do that for Boost as there are MANY THOUSANDS of files.

**Comment by Thunder07** [comment]
@Orphis this issue here though, that the git source is missing few files so it won't compile(I think this was by design to prevent git builds, not sure why) , so the release source has to be downloaded like this, or the entire thing has to be included (aka submodule won't work).

**Comment by Orphis** [comment]
Just copy the required sources in there, just like every other project do.

**Comment by Thunder07** [comment]
those files are generated as per OpenGL release or something akin to that, so I'm not sure how well will those copies work across Glew versions...
we can always generate them with the provided script, but that requires perl, while it won't be much of an issue for unix users, perhaps it will be for windows, to add to that Im certain Jean doesn't want nightly(?) version of glew being compiled(call me what you'd like, but i tend to pull the last updates from that branch), to add to that Glew maintainer recommends using releases as oppose to GitHub builds for obvious reasons.

So, I'm for downloading the source or adding it directly to repo. you're for copying the files, so I'll let jean decide this one

**Comment by Orphis** [comment]
Every project has a copy of the generated files. 
The only reason to regenerate them is when you want to use new OpenGL features that are just released, which is quite rare. Really, there's no downside there.

**Comment by Thunder07** [comment]
I know Jean has been specific with his OpenGL implementation to make sure he has compatibility between OpenGL and ES, so if thats the case, feature changes shouldn't be much of an issue.

anyways, as mentioned above, this will be close to the final draft of the script if you have any suggestions o improve it.

**Comment by jpd002** [comment]
@Thunder07 Thanks! I'll give your scripts a go whenever I get the chance!

As for GLEW, I think the generated files should be added as a dependency. As @Orphis mentioned, there's not much use for regenerating the files every time we build.

**Comment by Thunder07** [comment]
in the latest set of commits in branches `final_cmake`I've added glew into the dependencies, I've initially edited the CMakeLists.txt file to remove the install instruction from it, but decided to carry on and 'minimilise' it. 'I've also tested the setup with [travis](https://travis-ci.org/Thunder07/Play-/builds/191789677) and [appvoyer](https://ci.appveyor.com/project/Thunder07/play/build/1.0.9).

instead of re posting the setup instructions, I've update the post [above](https://github.com/jpd002/Play-/issues/506#issuecomment-271191392) or you can look at appvoyer.yml&.travis.sh [here](https://github.com/Thunder07/Play-/tree/final_cmake_travis), the instructions shouldnt really change from now on forward... beside perhaps the location of CMakeLists.txt

**Comment by jpd002** [comment]
I've been able to build the Win32 version of the emulator with the latest version of the CMake scripts. I'm going to try building for other platforms, but I think the current setup works great, even though there's still minor things that need to be tweaked.

How would you like to proceed for that next step? I think it might be interesting to open PRs so we can discuss those tweaks.

**Comment by Thunder07** [comment]
I've opened a new PR  for each repo. But before merging, I'd like to add a commit to fix the Travis and appvoyer build and perhaps even move the cmakelist out of build_unix directories and either into root or cmake folder since it's not Unix specific anymore.

**Comment by jpd002** [comment]
The `build_unix` could be renamed to `build_cmake`, but I wouldn't place the build files in the root directory because I prefer to keep all "build" related stuff in a separate folder.

**Comment by Thunder07** [comment]
to comment more about @Orphis comment of copying everything to `Play!`, I think perhaps having the other repos as submodules of this repo isn't a bad thing `Play/extern/_REPOS_`, its more appropriate than having a separate repo just for cloning(I know I was part of that initial suggestion :p ).

it has the advantage of being able to to trigger new builds with appropriate commit to update submodules checkout, as oppose to the current method were everything is checked out to master. since this is the main repo, we can get ride of `Play-Build.

**Comment by jpd002** [comment]
Yeah, I agree it would make sense to have the dependencies in a sub folder. My only worry is about external repos that reference the same external repos. 

For example, `Play` and `CodeGen` both need `Framework` to build, so, they should both have a `extern/Framework` submodule in their tree. But since `Play` needs `CodeGen` to build, would that mean you would get the `Framework` repo twice when cloning the `Play` repo (in `Play/extern/Framework` and `Play/extern/CodeGen/extern/Framework`)? 

There's certainly a nice way to organize that and I'm open to suggestions.

**Comment by Thunder07** [comment]
I see this working one of 2 ways
1) prioritising `Play!` dependencies (`extern`)
aka move `CodeGen`,`Framework`, `Dependencies`  into `Play/extern`, and `Framework` can refer to `CodeGen` as you do now. this method requires less work, the only change will be done in `Play` repo (`CMakeLists.txt` + submodules)
2) setup the project pasted on hierarchy of dependencies, since `Framework` refers to `CodeGen` 1st, it should contain it, and we can refer to it through from `Play`, through `CodeGen` extern. More work, change both `Framework` & `Play` (not that the changes are any harder)

I'm against 2, since `Play`'s dependency on `Framework` isn't directly dependent on `CodeGen`.
so, if you want to trigger and update due to `Framework` update, you'd need commit it in `CodeGen` 1st, then commit `CodeGen` of `Framework` for it to trigger. (you can use the same argument against `Play`, since if you use 1, you won't be able to trigger `Framework` update if you changed something in `CodeGen`, but as of this moment, you don't have build bots setup for `Framework`)

also, having to copy of `Framework` won't work with CMake, as it doesn't allow to build the same library (with the same name) twice. but can avoid this with `if (NOT TARGET Framework)`, but that comes another issue of included header, so imagine a hypothetical situation, where the user updated `Framework` in `Play!`, but not in `CodeGen`, and so `CodeGen` would use out of date headers while `Play` won't, no the compiled library is based on `CodeGen`'s `Framework`, yet `Play`'s `Framework` include are updated, thus you could potentially get a mismatch in header leading to linking or missing headers issue in `Play`

**Comment by Thunder07** [comment]
@jpd002 quick note on the upgrade to boost 1.64, to avoid you scratching your head trying to figure it out, it's something I've noticed when transitioning to `boost-cmake`, you probably noticed the linux build failure by now. thats because trusty is using `cmake 2.8` which is still using an old copy of `FindBZip2.cmake`, [this](https://github.com/Kitware/CMake/blob/23876eda9f1d4b5725407aef7d218fc3ce4113cb/Modules/FindBZip2.cmake) is the most likely version.
the 2 solutions are, ask/send PR to @Orphis to see if he can use library variable as as oppose to alias names in his modules.. or download cmake manually or backport repo... or revert back i guess.

**Comment by jpd002** [comment]
Thanks for the tips! I've pushed the fix on master and it should work properly now.

**Comment by Orphis** [comment]
Seriously, use a newer version of CMake.
Then, pre 3.0 the full version number is important, each 2.8.X version had new features.

As for Trusty, it's an old distribution now. The LTS aspect is good for servers, not so much for user's machines. So install a backport or the new version manually, it's not a big deal.

**Comment by Thunder07** [comment]
ummm.. trying telling that to Travis.... either way, it's been done already.

**Comment by Orphis** [comment]
As I said, it's good for servers, and if you want anything more recent, feel free to install it yourself, as you should anyway on all platforms or Linux distributions.

**Comment by LoungeKatt** [comment]
TL;DR @Orphis on iOS, I was using https://github.com/danoli3/ofxiOSBoost until cmake. I stopped building iOS because I rarely run it on iOS and the new configuration is a long list of macports / homebrew installation and configuration steps without building for it.

**Comment by jpd002** [comment]
Closing this since this is quite old. Feel free to open new issues to discuss specific problems with the build system.
